demos/jobs/jobs_demo_mosquitto/jobs_demo_mosquitto.c

```diff
@@ -0,0 +1,1264 @@
+            { "help",   no_argument,       NULL, '?' },
+            { NULL,     0,                 NULL, 0   }
+        };
+
+        c = getopt_long( argc, argv, "n:h:p:f:d:c:k:?",
+                         long_options, &option_index );
+
+        if( c == -1 )
+        {
+            break;
+        }
+
+        switch( c )
+        {
+            case 'n':
+                h->name = optarg;
+                h->nameLength = strlen( optarg );
+                break;
+
+            case 'h':
+                h->host = optarg;
+                break;
+
+            case 'p':
+                x = strtol( optarg, NULL, 0 );
+
+                if( ( x > 0 ) && ( x <= 0xFFFF ) )
+                {
+                    h->port = x;
+                }
+                else
+                {
+                    ret = false;
+                    warnx( "bad port value: %s", optarg );
+                }
+
+                break;
+
+            case 'f':
+                h->cafile = optarg;
+                h->capath = NULL;
+                break;
+
+            case 'd':
+                h->capath = optarg;
+                break;
+
+            case 'c':
+                h->cert = optarg;
+                break;
+
+            case 'k':
+                h->key = optarg;
+                break;
+
+            case '?':
+            default:
+                usage( argv[ 0 ] );
+        }
+    }
+
+    if( optind < argc )
+    {
+        ret = false;
+        usage( argv[ 0 ] );
+    }
+
+    if( ret == true )
+    {
+        ret = requiredArgs( h );
+    }
+
+    return ret;
+}
+
+/*-----------------------------------------------------------*/
+
+static void on_connect( struct mosquitto * m,
+                        void * p,
+                        int rc )
+{
+    handle_t * h = p;
+
+    assert( h != NULL );
+
+    h->connectError = rc;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool connect( handle_t * h )
+{
+    int ret = MOSQ_ERR_SUCCESS;
+    size_t i;
+
+    assert( h != NULL );
+    assert( h->m != NULL );
+    assert( h->connectError == -1 );
+
+    if( h->port == 443 )
+    {
+        #if ( LIBMOSQUITTO_VERSION_NUMBER >= 1006000 )
+            ret = mosquitto_string_option( h->m, MOSQ_OPT_TLS_ALPN, ALPN_NAME );
+        #else
+            warnx( "ALPN (port 443) is not supported by libmosquitto before version 1.6" );
+            ret = MOSQ_ERR_INVAL;
+        #endif
+    }
+
+    if( ret == MOSQ_ERR_SUCCESS )
+    {
+        ret = mosquitto_tls_set( h->m, h->cafile, h->capath, h->cert, h->key, NULL );
+    }
+
+    if( ret == MOSQ_ERR_SUCCESS )
+    {
+        info( "Connecting to %s, port %d.", h->host, h->port );
+        ret = mosquitto_connect( h->m, h->host, h->port, MQTT_KEEP_ALIVE );
+    }
+
+    /* expect the on_connect() callback to update h->connectError */
+    for( i = 0; ( i < MAX_LOOPS ) &&
+         ( ret == MOSQ_ERR_SUCCESS ) &&
+         ( h->connectError == -1 ); i++ )
```

Comment:
So if connection fails, that seems to be counted as fatal. Could we possible have backoff logic here? Although, I've never really had connection issues with AWS IoT to be honest.

Reply:
I want this to be a command-line utility like wget or curl.  A fast exit on failure is what you'd expect with these tools.  Should someone want to use this like a long running agent, I've given a suitable formula for use with cron in the usage statement.